LLM Scheduler

mahyarsamani opened the pull request.

mahyarsamani committed:
Initial Commit.

mahyarsamani committed:
Changing queue structure, Added Round Robin arbitration scheme,
Updated retry functionalities, Used better programming practices.
Still has bugs (injection rate > 100 breaks sim), scheduling
frequency needs to be fixed.

powerjg commented on src/mem/mem_scheduler.hh:
```diff
@@ -0,0 +1,362 @@
+/*
+ * Copyright (c) 2017 Jason Lowe-Power
```
Needs to be fixed

powerjg commented on src/mem/mem_scheduler.hh:
```diff
@@ -0,0 +1,362 @@
+#include "base/addr_range_map.hh"
+
+/**
+ * A very simple memory object. Current implementation doesn't even cache
+ * anything it just forwards requests and responses.
+ * This memobj is fully blocking (not non-blocking). Only a single request can
+ * be outstanding at a time.
```
Probably should update :)

powerjg replied:
Please update all of the comments.

powerjg commented on src/mem/mem_scheduler.hh:
```diff
@@ -0,0 +1,362 @@
+  private:
+
+    /**
+     * Port on the CPU-side that receives requests.
+     * Mostly just forwards requests to the owner.
+     * Part of a vector of ports. One for each CPU port (e.g., data, inst)
```
Reviewing for me, and understanding for you in a few months, is really hard when there's comments that don't match the code.

powerjg commented on src/mem/mem_scheduler.hh:
```diff
@@ -0,0 +1,362 @@
+     * Part of a vector of ports. One for each CPU port (e.g., data, inst)
+     */
+
+    struct RequestQueue{
```
General comments on this struct:

1. The style needs to be fixed. We use 4 spaces to indent.
2. All of these members need to have comments to explain what they are.
3. I think this is complex enough that it should be a class, not a struct. I.e., some of the members (most of them?) should be private.
4. Does it make sense for this to be part of the CPU port???

powerjg replied:
Ping...

powerjg replied:
ping...

Well, at this point, the `RequestQueue` has become too complicated to implement in a header. You should put the implementation in the cc file and make it a `class` not a `struct`

powerjg commented on src/mem/mem_scheduler.hh:
```diff
@@ -0,0 +1,362 @@
+         * Constructor. Just calls the superclass constructor.
+         */
+        MemSidePort(const std::string& name, PortID portId, MemScheduler *owner) :
+            RequestPort(name, owner), owner(owner), _blocked(false), blockedPacket(nullptr), _portId(portId)
```
line length... How are you bypassing gem5's style checker?! ;)

powerjg commented on src/mem/mem_scheduler.hh:
```diff
@@ -0,0 +1,362 @@
+        void sendPacket(PacketPtr pkt);
+
+        bool blocked(){
+          return _blocked;
```
Like so many places... this should be 4 spaces.

powerjg replied:
Ping

powerjg commented on src/mem/mem_scheduler.cc:
```diff
@@ -0,0 +1,402 @@
+                                    "should be non-zero");
+    for (uint32_t i = 0; i < nCpuPorts; ++i){
+        cpuPorts.emplace_back(name() + ".cpu_side" + std::to_string(i), i, this);
+        requestQueues.push_back(RequestQueue(readBufferSize, writeBufferSize, i));
```
Why push_back and not emplace? It's probably functionally the same, but push_back is creating a copy.

powerjg reviewed (commented):
Looks pretty good. I have a mix of comments on the modeling aspect and code style. 

Overall, you've done a very good job! However, I want to take the chance to reiterate something that's important.

One of the values of this group is reproducible and extensible scientific contributions. One aspect of this is making the code you work on open source and/or contributing upstream. This is not possible if you do not follow the style guidelines. I suggest just getting in the habit of *never* writing any code that doesn't follow the style of the code around it. It takes a bit to learn the style, but it shouldn't take you any longer to write the actual code. By doing it when you're writing the code the first time, you save the time of the code reviewers asking you to make updates.

The same goes with writing detailed comments. However, with comments it's not just for your reviewer, but for **you**. When working on a research project, *you* will be working on this code for at a minimum 1 year. You won't be constantly modifying every line. You must write good documentation so *you* can remember why you did something 6, 12, or more, months ago.

mahyarsamani committed:
Fixed bug on scheduling nextReqEvent. Style needs to be fixed.

marjanfariborz committed:
Adding unified queue and a simple arbiteration between read and write queue

mahyarsamani committed:
Updated queueing scheme to support both unified and separated read
and write queues.

marjanfariborz committed:
Bug fixed: sendRetry unified queue and setting write threshold to service write requests

mahyarsamani committed:
Added stats for readReqs, writeReqs, number of arbitrations.

powerjg commented on src/mem/mem_scheduler.cc:
```diff
@@ -0,0 +1,474 @@
+MemScheduler::MemSchedulerStat::regStats(){
+    avgRQDelay.precision(2);
+    avgWQDelay.precision(2);
+    avgRQDelay = totalRQDelay / readReqs;
+    avgWQDelay = totalWQDelay / writeReqs;
```
This can be done in the ADD_STAT macro

powerjg replied:
ping

mahyarsamani committed:
Started implementing optimal arbitration scheme, breaks code.

marjanfariborz committed:
Adding Memscheduler response delay

marjanfariborz committed:
Adding more statistics + updating the bank specifications

mahyarsamani committed:
Added a bunch of dprintfs to print timings in dram.

mahyarsamani committed:
Added more dprintfs.

marjanfariborz committed:
Fixing the timing parameters in LLM

mahyarsamani committed:
Fixed bugs in queue arbitration.

mahyarsamani committed:
Fixed typo.

mahyarsamani committed:
Updated linear_gen to create strided access, random_gen changed to
create uniform random pattern, updated mem_scheduler stats.

powerjg commented on src/cpu/testers/traffic_gen/linear_gen.cc:
```diff
@@ -71,7 +71,7 @@ LinearGen::getNextPacket()
 
     // increment the address
-    nextAddr += blocksize;
+    nextAddr += blocksize * 16;
```
Why this change?

powerjg replied:
Ping

powerjg commented on src/mem/DRAMInterface.py:
```diff
@@ -1414,3 +1414,184 @@ class LPDDR5_6400_1x16_8B_BL32(LPDDR5_6400_1x16_BG_BL32):
+
+class LLM2(DRAMInterface):
+
+    beats_per_clock = 16
```
I still think it's very important for you all to address the comments on this file.

powerjg commented on src/mem/MemScheduler.py:
```diff
@@ -0,0 +1,46 @@
+# -*- coding: utf-8 -*-
+# Copyright (c) 2017 Jason Lowe-Power
```
Still needs to be fixed.

powerjg commented on src/mem/MemScheduler.py:
```diff
@@ -0,0 +1,46 @@
+    resp_buffer_size = Param.Int(8, "Size of the internal "
+                                    "response buffer")
+    service_write_threshold = Param.Percent(60, "Threshold to force writes")
+    unified_queue = Param.Bool(False, "Use same queue for reads and "
```
If this is true, what is the size of the queue? The read buffer size, write buffer size, the sum?

powerjg commented on src/mem/mem_interface.cc:
```diff
@@ -289,11 +289,13 @@ DRAMInterface::activateBank(Rank& rank_ref, Bank& bank_ref,
     // if not, shift to next burst window
     Tick act_at;
-    if (twoCycleActivate)
+    if (twoCycleActivate){
```
Ugh... this is why you should always use curly braces! (I'm just pointing out the poor coding practice that was already there.)

You're missing a space. (Below as well)

powerjg commented on src/mem/mem_interface.cc:
```diff
@@ -504,15 +510,17 @@ DRAMInterface::doBurstAccess(MemPacket* mem_pkt, Tick next_burst_at,
     if (dataClockSync && ((cmd_at - rank_ref.lastBurstTick) > clkResyncDelay))
         cmd_at = ctrl->verifyMultiCmd(cmd_at, maxCommandsPerWindow, tCK);
-    else
+    else{
```
space
```suggestion
    else {
```

powerjg commented on src/mem/mem_interface.hh:
```diff
@@ -924,7 +924,7 @@ class DRAMInterface : public MemInterface
      * @return time to offset next command
      */
-    Tick commandOffset() const override { return (tRP + tRCD); }
+    Tick commandOffset() const override { return (tRCD); }
```
Is this a bug?!?!

Why this change?

powerjg replied:
Ping

powerjg commented on src/mem/mem_scheduler.cc:
```diff
@@ -0,0 +1,634 @@
+    }
+    if (!nextReqEvent.scheduled()){
+        DPRINTF(MemScheduler, "handleRequest: Scheduling nextReqEvent in handleRequest\n");
+        schedule(nextReqEvent, curTick() + 100);
```
What is this 100???

powerjg commented on src/mem/mem_scheduler.cc:
```diff
@@ -0,0 +1,634 @@
+    }
+
+    stats.totalArbitrations++;
+    DPRINTF(MemScheduler, "processNextReqEvent: Least recently visited queue found, cpuPortId = %d, (timesChecked = %d) == (minChecked = %d), readQueue.size = %d, writeQueue.size = %d, emptyRead = %d, serviceWrite = %d\n", queue->cpuPortId, queue->timesChecked, minCheck, queue->readQueue.size(), queue->writeQueue.size(), queue->emptyRead(), queue->serviceWrite());
```
This line is ridiculous. 80 chars might be small, but there has to be some limit. You really should be following gem5's style guideline to a "T" even if you're working on research.

How can anyone read/understand this line?

powerjg commented on src/mem/mem_scheduler.cc:
```diff
@@ -0,0 +1,634 @@
+        DPRINTF(MemScheduler, "processNextReqEvent: Servicing write request now\n");
+        pkt = queue->writeQueue.front();
+    }
+    AddrRange addr_range = pkt->getAddrRange();
```
Why range and not the address?

powerjg commented on src/mem/mem_scheduler.cc:
```diff
@@ -0,0 +1,634 @@
+    DPRINTF(MemScheduler, "processNextReqEvent: pkt->addr_range: %s\n", addr_range.to_string());
+    PortID memPortId = memPortMap.contains(addr_range)->second;
+    DPRINTF(MemScheduler, "processNextReqEvent: Looked up outgoing routing table for MemSidePort PortID: %d\n", memPortId);
+    auto memPort = find_if(memPorts.begin(), memPorts.end(), [memPortId](MemSidePort &obj){return obj.portId() == memPortId;});
```
It looks like memPorts should be a map instead of whatever it is. `find_if` is a strange thing to use here.

powerjg commented on src/mem/mem_scheduler.cc:
```diff
@@ -0,0 +1,634 @@
+    }
+    auto pick = arbitrate(visited);
+    
+    if (pick == NULL){
```
use `nullptr` instead of `NULL`

powerjg reviewed (changes requested):
I *advise* you to take an hour or two and go through all of my comments here and address them. I don't think it will take that long and I think it's important.

Generally, in mem_scheduler.cc I'm having a really hard time reading and reviewing the code because you have completely ignored the style guide. It's *very important* to follow the style guide of whatever code you're working on. As a researcher (and especially in this group) you are *never* writing code only for yourself. You're always writing code that others should be able to understand and contribute to. You must follow the style guide to make it feasible for others to read and understand and contribute to your code.

marjanfariborz committed:
mem: Updating the style and adding more comments.

Coding style has been changed. More comments are added to the
memory interface.

Change-Id: Id72c6c5eb5268623b432640ddb4ef27d070e90bf

mahyarsamani committed:
mem: Fixing bugs when sending write requests

Moving sendPacket below the book keeping section of the code that
rely on the information acquired by dereferencing a PacketPtr.
Updating DRAMInterface params for LLM to reflect timings for
writes and reads together.

Change-Id: Ibb73d5325acb710f3af27741a6d33a86f4bae186

powerjg commented on src/cpu/testers/traffic_gen/random_gen.cc:
```diff
@@ -61,10 +61,11 @@ RandomGen::getNextPacket()
-    Addr addr = random_mt.random(startAddr, endAddr - 1);
-
+    Addr addr = random_mt.random(startAddr / blocksize, endAddr / blocksize - 1);
+    DPRINTF(TrafficGen, "startAddr: %u, endAddr: %u\n", startAddr, endAddr);
     // round down to start address of block
-    addr -= addr % blocksize;
+    // addr -= addr % blocksize;
+    addr *= blocksize;
```
Probably should revert this.

powerjg commented on src/mem/DRAMInterface.py:
```diff
@@ -1414,3 +1414,97 @@ class LPDDR5_6400_1x16_8B_BL32(LPDDR5_6400_1x16_BG_BL32):
     tRRD_L = "0ns"
     tWTR_L = "0ns"
+
+class LLM2(DRAMInterface):
```
lol, I like that you're starting with "2". Who needs a 1.0?!

powerjg commented on src/mem/DRAMInterface.py:
```diff
@@ -1414,3 +1414,97 @@ class LPDDR5_6400_1x16_8B_BL32(LPDDR5_6400_1x16_BG_BL32):
+
+class LLM2(DRAMInterface):
+
+    beats_per_clock = 64
```
Can you add a comment here that says "SERDES running at 32 GHz" assuming that is what `64` beats means.

powerjg commented on src/mem/DRAMInterface.py:
```diff
@@ -1414,3 +1414,97 @@ class LPDDR5_6400_1x16_8B_BL32(LPDDR5_6400_1x16_BG_BL32):
+    # memory access latency
+    page_policy = 'close'
+
+    # Total number of pins are 1024
+    # There are 1 pin/channel
+    # Atom size =  32Byte , BL= 64/1 (Bytes)
+    burst_length = 512
```
This confuses me a bit.  This is the number of beats needed for 64 bytes of data? Since it's fully serialized, you need 64*8=512 beats?

Putting this in the comment would be helpful!

powerjg commented on src/mem/DRAMInterface.py:
```diff
@@ -1414,3 +1414,97 @@ class LPDDR5_6400_1x16_8B_BL32(LPDDR5_6400_1x16_BG_BL32):
     tCCD_L = "0ns"
     tRRD_L = "0ns"
     tWTR_L = "0ns"
+
```
This should be in a new file (LLMInterface.py)

In fact, there's no need to put this in the `src/` directory either. All of this is configurable at runtime.

powerjg commented on src/mem/DRAMInterface.py:
```diff
@@ -1414,3 +1414,184 @@ class LPDDR5_6400_1x16_8B_BL32(LPDDR5_6400_1x16_BG_BL32):
+    tBURST_MAX = '4ns'
+    # BL2 and BL4 supported, default to BL4
+    # DDR @ 500 MHz means 4 * 2ns / 2 = 4ns
+    tBURST = '4ns'
```
512 beats / 64 beats per clock = 8 clock cycles @ 500 MHz is 16ns, right??

Please update the comments above to make sense with your device. Don't quote DDR unless you have to.

powerjg commented on src/mem/DRAMInterface.py:
```diff
@@ -1414,3 +1414,97 @@ class LPDDR5_6400_1x16_8B_BL32(LPDDR5_6400_1x16_BG_BL32):
+
+
+    tWR = '18ns'
+    # wewill mask this we consecutive requests
```
I don't know what this means.

powerjg commented on src/mem/DRAMInterface.py:
```diff
@@ -1414,3 +1414,97 @@ class LPDDR5_6400_1x16_8B_BL32(LPDDR5_6400_1x16_BG_BL32):
+    tWTR = '0ns'
+    tWTR_L = '0ns'
+
+    # start with 2 cycles turnaround, similar to other memory classes
+    # could be more with variations across the stack
+    tRTW = '0ns'
```
This comment doesn't match the value.

powerjg commented on src/mem/DRAMInterface.py:
```diff
@@ -1414,3 +1414,97 @@ class LPDDR5_6400_1x16_8B_BL32(LPDDR5_6400_1x16_BG_BL32):
+    # single rank device, set to 0
+    tCS = '0ns'
+
+    # from MemCon example, tRRD is 4ns with 2ns tCK
```
What is "MemCon example"?

powerjg commented on src/mem/mem_interface.cc:
```diff
@@ -683,6 +699,10 @@ DRAMInterface::doBurstAccess(MemPacket* mem_pkt, Tick next_burst_at,
         stats.totQLat += cmd_at - mem_pkt->entryTime;
         stats.totBusLat += tBURST;
+
+        if (mem_pkt->readyTime - mem_pkt->entryTime > 24){
+            stats.bankConflicts++;
+        }
```
What is this magic 24 number?

powerjg replied:
ping

powerjg commented on src/mem/mem_scheduler.hh:
```diff
@@ -0,0 +1,398 @@
+/*
+ * Copyright (c) 2020 UC regents
```
```suggestion
 * Copyright (c) 2020 The Regents of the University of California.
```

powerjg commented on src/mem/mem_scheduler.hh:
```diff
@@ -0,0 +1,398 @@
+#include "mem/port.hh"
+#include "params/MemScheduler.hh"
+
+// #include "sim/sim_object.hh"
```
Don't just comment things out.

Before you push any commits, run `git diff` or `git log -n1 --patch` or `git diff --cached` and read every single line that you've changed. Look for style things like whether there's commented out code.

powerjg commented on src/mem/mem_scheduler.hh:
```diff
@@ -0,0 +1,398 @@
+#include "sim/clocked_object.hh"
+
+/**
+ * A very simple memory object. This object can be connected to multiple memory
```
"A very simple memory object"

Does this really describe this class? Is this not a "memory scheduler to be used with LLM" or something like that?

powerjg commented on src/mem/mem_scheduler.hh:
```diff
@@ -0,0 +1,398 @@
+    class MemSidePort : public RequestPort
+    {
+      private:
+          /// The object that owns this object (MemScheduler)
+          MemScheduler *owner;
+          bool _blocked;
+          /// If we tried to send a packet and it was blocked, store it here
+          PacketPtr blockedPacket;
+          const PortID _portId;
+      public:
+          /**
+          * Constructor. Just calls the superclass constructor.
+          */
+          MemSidePort(const std::string& name, PortID portId,
+                      MemScheduler *owner) :
+              RequestPort(name, owner), owner(owner), _blocked(false),
+                          blockedPacket(nullptr), _portId(portId)
+          {}
```
Indented too many times.

powerjg commented on src/mem/mem_scheduler.hh:
```diff
@@ -0,0 +1,398 @@
+    };
+
+    /**
+     * Handle the request from the CPU side
```
Can you describe at a high level what is supposed to happen on a request?

powerjg commented on src/mem/mem_scheduler.hh:
```diff
@@ -0,0 +1,398 @@
+    /// Instantiation of the memory-side port
+    std::vector<MemSidePort> memPorts;
+
+    RequestQueue* arbitrate(std::map<PortID, bool> visited);
```
Should have some comments.

powerjg commented on src/mem/mem_scheduler.hh:
```diff
@@ -0,0 +1,398 @@
+    std::unordered_map<PacketPtr, Tick> respEntryTimes;
+    struct MemSchedulerStat : public Stats::Group
+    {
+      MemSchedulerStat(MemScheduler *parent);
```
four spaces

powerjg commented on src/mem/mem_scheduler.cc:
```diff
@@ -0,0 +1,778 @@
+    for (uint32_t i = 0; i < nCpuPorts; ++i){
+        cpuPorts.emplace_back(name() + ".cpu_side" + std::to_string(i),
+                              i, this);
+        requestQueues.push_back(RequestQueue(readBufferSize, writeBufferSize,
```
probably want `emplace` instead of `push`. Push probably does a deep copy.

powerjg commented on src/mem/mem_scheduler.cc:
```diff
@@ -0,0 +1,778 @@
+                              i, this);
+        requestQueues.push_back(RequestQueue(readBufferSize, writeBufferSize,
+                            params.service_write_threshold, unifiedQueue, i));
+        responseQueues.push_back(ResponseQueue(respBufferSize, i));
```
ditto

powerjg commented:
Is there any reason for me to look at this again? There are still *many* unresolved comments from previous reviews.

mahyarsamani commented:
> Is there any reason for me to look at this again? There are still _many_ unresolved comments from previous reviews.

No worries, I'm actually doing a lot of cleaning up today, you can expect to see a lot of these comments resolved tomorrow!

marjanfariborz committed:
mem: Removing unnecessary changes.

Changes to files that were unnecessary to implement LLMScheduler
have been removed.

marjanfariborz committed:
mem: Separating LLMInterface

A separate file has created for LLM and describes the timing
constraints of LLM design.

Change-Id: I623d6f50b97f44c938f72843e2755c442690e043

marjanfariborz committed:
mem: Adding functional and atomic interface

This patch adds functional and atomic interfaces
which are essential for FS runs.

Change-Id: Ib77a4cf0dce377e4bcca974814f059a0d94311d7

powerjg commented on src/mem/LLMInterface.py:
```diff
@@ -0,0 +1,121 @@
+# Copyright (c) 2020 The Regents of the University of California.
```
This file looks great! Thanks for all of the updates! A small thing ;)
 
```suggestion
# Copyright (c) 2021 The Regents of the University of California.
```

powerjg commented on src/mem/mem_scheduler.hh:
```diff
@@ -0,0 +1,396 @@
+      sendWriteRetry(false){}
+    };
+
+    struct ResponseQueue{
```
Would it make sense for this to be a `PacketQueue`?

powerjg commented:
I skipped the `MemScheduler` as it doesn't seem like my previous comments have been addressed.

mahyarsamani committed:
mem: Adding infinite queuing option.

mahyarsamani committed:
mem: Updating LLMInterface, fixing arbitration scheme.

Change-Id: Ib3570ef2ea7350dc748d23536d67587a75b8645e

marjanfariborz committed:
mem: Adding interface params for FGDRAM.

Change-Id: I6c4f9fa279c50e43a1669956a27c111d6c7eb0f2

marjanfariborz committed:
base: Increasing the total number of possible ranges in a system.

Change-Id: I0ac7febd95eb49acffe06a751a20fa80ecd87231

mahyarsamani committed:
mem: Updating schedule event in MemScheduler.

mahyarsamani committed:
mem: Updating params for FGDRAM

Change-Id: If29450c6c6879e430a5a58d1f02cb8b8d4eb0480

marjanfariborz committed:
Suppoting SALP in mem_iterface
